llap-client/src/test/org/apache/hadoop/hive/llap/registry/impl/TestLlapZookeeperRegistryImpl.java

```diff
@@ -0,0 +1,124 @@
+    server.start();
+
+    curatorFramework = CuratorFrameworkFactory.
+            builder().
+            connectString(server.getConnectString()).
+            sessionTimeoutMs(1000).
+            retryPolicy(new RetryOneTime(1000)).
+            build();
+    curatorFramework.start();
+
+    trySetMock(registry, CuratorFramework.class, curatorFramework);
+
+  }
+
+  @After
+  public void tearDown() throws IOException {
+    curatorFramework.close();
+    server.stop();
+  }
+
+  @Test
+  public void testRegister() throws Exception {
+    // Given
+    int expectedExecutorCount = HiveConf.getIntVar(mockConf, HiveConf.ConfVars.LLAP_DAEMON_NUM_EXECUTORS);
+    int expectedQueueSize = HiveConf.getIntVar(mockConf, HiveConf.ConfVars.LLAP_DAEMON_TASK_SCHEDULER_WAIT_QUEUE_SIZE);
+
+    // When
+    registry.register();
+    ServiceInstanceSet<LlapServiceInstance> serviceInstanceSet =
+      registry.getInstances("LLAP", 1000);
+
+    // Then
+    Collection<LlapServiceInstance> llaps = serviceInstanceSet.getAll();
+    assertEquals(1, llaps.size());
+    LlapServiceInstance serviceInstance = llaps.iterator().next();
+    Map<String, String> attributes = serviceInstance.getProperties();
+
+    assertEquals(expectedQueueSize,
+      parseInt(attributes.get(LlapRegistryService.LLAP_DAEMON_TASK_SCHEDULER_ENABLED_WAIT_QUEUE_SIZE)));
+    assertEquals(expectedExecutorCount,
+      parseInt(attributes.get(LlapRegistryService.LLAP_DAEMON_NUM_ENABLED_EXECUTORS)));
+  }
+
+  @Test
+  public void testUpdate() throws Exception {
+    // Given
+    String expectedExecutorCount = "2";
+    String expectedQueueSize = "20";
+    Map<String, String> capacityValues = new HashMap<>(2);
+    capacityValues.put(LlapRegistryService.LLAP_DAEMON_NUM_ENABLED_EXECUTORS, expectedExecutorCount);
+    capacityValues.put(LlapRegistryService.LLAP_DAEMON_TASK_SCHEDULER_ENABLED_WAIT_QUEUE_SIZE, expectedQueueSize);
+
+    // When
+    registry.register();
+    registry.updateRegistration(capacityValues.entrySet());
+    ServiceInstanceSet<LlapServiceInstance> serviceInstanceSet =
+            registry.getInstances("LLAP", 1000);
+
+    // Then
+    Collection<LlapServiceInstance> llaps = serviceInstanceSet.getAll();
+    assertEquals(1, llaps.size());
+    LlapServiceInstance serviceInstance = llaps.iterator().next();
+    Map<String, String> attributes = serviceInstance.getProperties();
+
+    assertEquals(expectedQueueSize,
+            attributes.get(LlapRegistryService.LLAP_DAEMON_TASK_SCHEDULER_ENABLED_WAIT_QUEUE_SIZE));
+    assertEquals(expectedExecutorCount,
+            attributes.get(LlapRegistryService.LLAP_DAEMON_NUM_ENABLED_EXECUTORS));
+  }
+
+  static <T> void trySetMock(Object o, Class<T> clazz, T mock) {
+    List<InstanceField> instanceFields = Fields
+      .allDeclaredFieldsOf(o)
+      .filter(instanceField -> !clazz.isAssignableFrom(instanceField.jdkField().getType()))
+      .instanceFields();
+    if (instanceFields.size() != 1) {
```

Comment:
Wouldn't it be easier to specify the field by name, rather than by class? In that case you wouldn't have issues on two fields with same type....

Reply:
It would be easier, but I think this way it's more robust.